File: FSUIKit/Classes/FSView.m
////
////  FSView.m
////  FSBaseController
////
////  Created by FudonFuchina on 2019/5/12.
////
//
//#import "FSView.h"
//
//@implementation FSView
//
//- (void)layoutSubviews {
//    [super layoutSubviews];
//    _tapBackView.frame = self.bounds;
//    _gradientLayer.frame = self.bounds;
//}
//
//- (CAGradientLayer *)gradientLayer {
//    if (!_gradientLayer) {
//        _gradientLayer = [CAGradientLayer layer];
//        _gradientLayer.frame = self.bounds;
//        [self.layer insertSublayer: _gradientLayer atIndex: 0];
//    }
//    return _gradientLayer;
//}
//
//- (instancetype)initWithFrame:(CGRect)frame {
//    self = [super initWithFrame:frame];
//    if (self) {
//        [self fs_add_tap_event_in_base_view];
//    }
//    return self;
//}
//
//- (void)fs_add_tap_event_in_base_view {
//    _tapBackView = [[UIView alloc] initWithFrame: self.bounds];
//    [self addSubview: _tapBackView];
//    
//    UITapGestureRecognizer *one = [[UITapGestureRecognizer alloc] initWithTarget: self action: @selector(_fs_tap_click_event:)];
//    [_tapBackView addGestureRecognizer: one];
//}
//
//- (void)_fs_tap_click_event:(UITapGestureRecognizer *)tap {
//    if (self.click) {
//        self.click(self);
//    }
//    
//    if (self.clickLocation) {
//        CGPoint p = [tap locationInView: _tapBackView];
//        self.clickLocation(self, p);
//    }
//}
//
//+ (FSView *)viewWithTheTag:(NSInteger)tag inView:(nonnull UIView *)inView {
//    for (FSView *sub in inView.subviews) {
//        if ([sub isKindOfClass:FSView.class]) {
//            if (sub.theTag == tag) {
//                return sub;
//            }
//        }
//    }
//    return nil;
//}
//
//- (void)dismiss {
//    [UIView animateWithDuration:.25 animations:^{
//        self.alpha = 0;
//    } completion:^(BOOL finished) {
//        [self removeFromSuperview];
//    }];
//}
//
///*
//// Only override drawRect: if you perform custom drawing.
//// An empty implementation adversely affects performance during animation.
//- (void)drawRect:(CGRect)rect {
//    // Drawing code
//}
//*/
//
//@end

File: FSUIKit/Classes/FSView.swift
//
//  FSViewS.swift
//  FSUIKit
//
//  Created by Dongdong Fu on 2026/1/29.
//

import Foundation

open class FSView: UIView {
    
    public var  tapBackView                             :   UIView?                             =   nil
    public var  click                                   :   ((FSView, CGPoint) ->  Void)?       =   nil
    
    public var  theTag                                  :   Int                                 =   0
    
    private var _gradientLayer                          :   CAGradientLayer?

    lazy var gradientLayer: CAGradientLayer = {
        let g = CAGradientLayer()
        g.frame = bounds
        self.layer.insertSublayer(g, at: 0)
        return g
    }()
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        
        if self.tapBackView != nil {
            self.tapBackView?.frame = bounds
        }
        
        if let gradientLayer = _gradientLayer {
            gradientLayer.frame = bounds
        }
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        
        self.fs_add_tap_event_in_base_view()
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func fs_add_tap_event_in_base_view() {
        tapBackView = UIView(frame: bounds)
        self.addSubview(tapBackView!)
        
        let one = UITapGestureRecognizer(target: self, action: #selector(_fs_tap_click_event(_:)))
        tapBackView!.addGestureRecognizer(one)
    }
    
    @objc func _fs_tap_click_event(_ tap: UITapGestureRecognizer) {
        
        if self.click != nil {
            let p = tap.location(in: tapBackView)
            self.click!(self, p)
        }
    }
    
    public static func viewWithTheTag(tag: Int, view: UIView) -> FSView? {
        
        for sub in view.subviews {
            if sub is FSView {
                let subs: FSView = sub as! FSView
                if subs.theTag == tag {
                    return subs
                }
            }
        }
        
        return nil
    }
    
    open func dismiss() {
        UIView.animate(withDuration: 0.25) {
            self.alpha = 0
        } completion: { f in
            self.removeFromSuperview()
        }
    }
}
